sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+   * auxiliary column "isFiltered"
+   * @param subqueryPlan plan before rewrite
+   * @param filteredId expression ID for the "isFiltered" column
+   */
+  private def addIsFiltered(subqueryPlan : LogicalPlan, filteredId : ExprId) : LogicalPlan = {
+    val isFilteredRef = AttributeReference("isFiltered", BooleanType)(exprId = filteredId)
+    val (topPart, aggNode) = splitSubquery(subqueryPlan)
+    var rewrittenQuery: LogicalPlan = null
+    if (topPart.size > 0 && topPart.head.isInstanceOf[Filter]) {
+      // Correlated subquery has a HAVING clause
+      // Rewrite the Filter into a Project that returns the value of the filtering predicate
+      val origFilter = topPart.head.asInstanceOf[Filter]
+      var topRemainder = topPart.tail
+      val newProjectList =
+        origFilter.output :+ Alias(origFilter.condition, "isFiltered")(exprId = filteredId)
+      val filterAsProject = Project(newProjectList, origFilter.child)
+
+      rewrittenQuery = filterAsProject
+      while (topRemainder.size > 0) {
+        rewrittenQuery = topRemainder.head match {
+          case Project(origList, _) => Project(origList :+ isFilteredRef, rewrittenQuery)
+          case SubqueryAlias(alias, _) => SubqueryAlias(alias, rewrittenQuery)
+        }
+        topRemainder = topRemainder.tail
+      }
+    } else {
+      // Correlated subquery without HAVING clause
+      // Add an additional Project that adds a constant value for "isFiltered"
+      rewrittenQuery = Project(subqueryPlan.output :+ Alias(Literal(false), "isFiltered")
+      (exprId = filteredId), subqueryPlan)
+    }
+    return rewrittenQuery
+  }
+
+  /**
    * Construct a new child plan by left joining the given subqueries to a base plan.
    */
   private def constructLeftJoins(
       child: LogicalPlan,
       subqueries: ArrayBuffer[ScalarSubquery]): LogicalPlan = {
     subqueries.foldLeft(child) {
       case (currentChild, ScalarSubquery(query, conditions, _)) =>
-        Project(
-          currentChild.output :+ query.output.head,
-          Join(currentChild, query, LeftOuter, conditions.reduceOption(And)))
+        val origOutput = query.output.head
+
+        val resultWithZeroTups = evalSubqueryOnZeroTups(query)
+        if (resultWithZeroTups.isEmpty) {
+          Project(
+            currentChild.output :+ origOutput,
+            Join(currentChild, query, LeftOuter, conditions.reduceOption(And)))
+        } else {
+          // Renumber the original output, because the outer query refers to its ID.
+          val newExprId = NamedExpression.newExprId
+          val renumberedQuery = query transformExpressions {
+            case a@Alias(c, n) if a.exprId == origOutput.exprId => Alias(c, n)(exprId = newExprId)
+          }
+
+          val filteredId = NamedExpression.newExprId
+          val isFilteredRef = AttributeReference("isFiltered", BooleanType)(exprId = filteredId)
+          val withIsFiltered = addIsFiltered(renumberedQuery, filteredId)
```

Comment:
Instead of turn the predicates in subquery into a projection, can we pull out the predicate in HAVING, then evaluate that after the join, so the result column will be.
```
Project(currentChild.output :+ If(predicate, agg, Literal(null)),
  Project(currentChild.output :+ Alias(Coalesce(Seq(aggvalRef, resultWithZeroTups)), "agg")())
```